File: templates/wasm-node-python/examples/math_nodes.py
"""
Math Nodes — Basic arithmetic and mathematical operations

Demonstrates creating multiple nodes for add, subtract, multiply,
divide, power, and clamp operations.
"""

from sdk import (
    Context,
    ExecutionResult,
    NodeDefinition,
    PinDefinition,
    PinType,
)


def _base_math_node(name: str, friendly_name: str, description: str) -> NodeDefinition:
    nd = NodeDefinition(name, friendly_name, description, "Math/Arithmetic")
    nd.add_pin(PinDefinition.input_exec("exec"))
    nd.add_pin(PinDefinition.input_pin("a", PinType.F64, default=0.0))
    nd.add_pin(PinDefinition.input_pin("b", PinType.F64, default=0.0))
    nd.add_pin(PinDefinition.output_exec("exec_out"))
    nd.add_pin(PinDefinition.output_pin("result", PinType.F64))
    return nd


def get_definitions() -> list[NodeDefinition]:
    nodes: list[NodeDefinition] = []

    nodes.append(_base_math_node("math_add_py", "Add", "Adds two numbers together"))
    nodes.append(_base_math_node("math_subtract_py", "Subtract", "Subtracts B from A"))
    nodes.append(_base_math_node("math_multiply_py", "Multiply", "Multiplies two numbers"))

    # Divide has an extra output
    divide = NodeDefinition("math_divide_py", "Divide", "Divides A by B", "Math/Arithmetic")
    divide.add_pin(PinDefinition.input_exec("exec"))
    divide.add_pin(PinDefinition.input_pin("a", PinType.F64, default=0.0))
    divide.add_pin(PinDefinition.input_pin("b", PinType.F64, default=1.0))
    divide.add_pin(PinDefinition.output_exec("exec_out"))
    divide.add_pin(PinDefinition.output_pin("result", PinType.F64))
    divide.add_pin(PinDefinition.output_pin("is_valid", PinType.BOOL))
    nodes.append(divide)

    # Power
    power = NodeDefinition("math_power_py", "Power", "Raises A to the power of B", "Math/Arithmetic")
    power.add_pin(PinDefinition.input_exec("exec"))
    power.add_pin(PinDefinition.input_pin("base", PinType.F64, default=0.0))
    power.add_pin(PinDefinition.input_pin("exponent", PinType.F64, default=1.0))
    power.add_pin(PinDefinition.output_exec("exec_out"))
    power.add_pin(PinDefinition.output_pin("result", PinType.F64))
    nodes.append(power)

    # Clamp
    clamp = NodeDefinition("math_clamp_py", "Clamp", "Clamps a value between min and max", "Math/Utility")
    clamp.add_pin(PinDefinition.input_exec("exec"))
    clamp.add_pin(PinDefinition.input_pin("value", PinType.F64, default=0.0))
    clamp.add_pin(PinDefinition.input_pin("min", PinType.F64, default=0.0))
    clamp.add_pin(PinDefinition.input_pin("max", PinType.F64, default=1.0))
    clamp.add_pin(PinDefinition.output_exec("exec_out"))
    clamp.add_pin(PinDefinition.output_pin("result", PinType.F64))
    nodes.append(clamp)

    return nodes


def run_add(ctx: Context) -> ExecutionResult:
    a = ctx.get_f64("a", 0.0)
    b = ctx.get_f64("b", 0.0)
    ctx.set_output("result", a + b)
    return ctx.success()


def run_subtract(ctx: Context) -> ExecutionResult:
    a = ctx.get_f64("a", 0.0)
    b = ctx.get_f64("b", 0.0)
    ctx.set_output("result", a - b)
    return ctx.success()


def run_multiply(ctx: Context) -> ExecutionResult:
    a = ctx.get_f64("a", 0.0)
    b = ctx.get_f64("b", 0.0)
    ctx.set_output("result", a * b)
    return ctx.success()


def run_divide(ctx: Context) -> ExecutionResult:
    a = ctx.get_f64("a", 0.0)
    b = ctx.get_f64("b", 1.0)

    if b == 0.0:
        ctx.set_output("result", 0.0)
        ctx.set_output("is_valid", False)
        ctx.warn("Division by zero")
    else:
        ctx.set_output("result", a / b)
        ctx.set_output("is_valid", True)

    return ctx.success()


def run_power(ctx: Context) -> ExecutionResult:
    base = ctx.get_f64("base", 0.0)
    exponent = ctx.get_f64("exponent", 1.0)
    ctx.set_output("result", base**exponent)
    return ctx.success()


def run_clamp(ctx: Context) -> ExecutionResult:
    value = ctx.get_f64("value", 0.0)
    min_val = ctx.get_f64("min", 0.0)
    max_val = ctx.get_f64("max", 1.0)
    ctx.set_output("result", max(min_val, min(max_val, value)))
    return ctx.success()


DISPATCH = {
    "math_add_py": run_add,
    "math_subtract_py": run_subtract,
    "math_multiply_py": run_multiply,
    "math_divide_py": run_divide,
    "math_power_py": run_power,
    "math_clamp_py": run_clamp,
}


def run(node_name: str, ctx: Context) -> ExecutionResult:
    handler = DISPATCH.get(node_name)
    if handler is None:
        return ctx.fail(f"Unknown node: {node_name}")
    return handler(ctx)

File: templates/wasm-node-assemblyscript/examples/math_nodes.ts
import {
	DataType,
	ExecutionResult,
	NodeDefinition,
	PinDefinition,
	Context,
	FlowNode,
	NodePackage,
} from "@flow-like/wasm-sdk/assembly/index";

export { alloc, dealloc, get_abi_version } from "@flow-like/wasm-sdk/assembly/index";

export class AddNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "math_add_as";
		def.friendly_name = "Add (AS)";
		def.description = "Adds two numbers together";
		def.category = "Math/Arithmetic";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("a", "A", "First number", DataType.F64).withDefaultF64(0.0));
		def.addPin(PinDefinition.input("b", "B", "Second number", DataType.F64).withDefaultF64(0.0));
		def.addPin(PinDefinition.output("exec_out", "Done", "Complete", DataType.Exec));
		def.addPin(PinDefinition.output("result", "Result", "Sum of A and B", DataType.F64));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		ctx.setF64("result", ctx.getF64("a") + ctx.getF64("b"));
		return ctx.success();
	}
}

export class SubtractNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "math_subtract_as";
		def.friendly_name = "Subtract (AS)";
		def.description = "Subtracts B from A";
		def.category = "Math/Arithmetic";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("a", "A", "First number", DataType.F64).withDefaultF64(0.0));
		def.addPin(PinDefinition.input("b", "B", "Second number", DataType.F64).withDefaultF64(0.0));
		def.addPin(PinDefinition.output("exec_out", "Done", "Complete", DataType.Exec));
		def.addPin(PinDefinition.output("result", "Result", "A minus B", DataType.F64));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		ctx.setF64("result", ctx.getF64("a") - ctx.getF64("b"));
		return ctx.success();
	}
}

export class MultiplyNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "math_multiply_as";
		def.friendly_name = "Multiply (AS)";
		def.description = "Multiplies two numbers";
		def.category = "Math/Arithmetic";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("a", "A", "First number", DataType.F64).withDefaultF64(1.0));
		def.addPin(PinDefinition.input("b", "B", "Second number", DataType.F64).withDefaultF64(1.0));
		def.addPin(PinDefinition.output("exec_out", "Done", "Complete", DataType.Exec));
		def.addPin(PinDefinition.output("result", "Result", "A times B", DataType.F64));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		ctx.setF64("result", ctx.getF64("a") * ctx.getF64("b"));
		return ctx.success();
	}
}

export class DivideNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "math_divide_as";
		def.friendly_name = "Divide (AS)";
		def.description = "Divides A by B";
		def.category = "Math/Arithmetic";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("a", "A", "Dividend", DataType.F64).withDefaultF64(0.0));
		def.addPin(PinDefinition.input("b", "B", "Divisor", DataType.F64).withDefaultF64(1.0));
		def.addPin(PinDefinition.output("exec_out", "Done", "Complete", DataType.Exec));
		def.addPin(PinDefinition.output("result", "Result", "A divided by B", DataType.F64));
		def.addPin(PinDefinition.output("is_valid", "Valid", "False if division by zero", DataType.Bool));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		const a = ctx.getF64("a");
		const b = ctx.getF64("b");

		if (b == 0.0) {
			ctx.setF64("result", 0.0);
			ctx.setBool("is_valid", false);
		} else {
			ctx.setF64("result", a / b);
			ctx.setBool("is_valid", true);
		}
		return ctx.success();
	}
}

export class ClampNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "math_clamp_as";
		def.friendly_name = "Clamp (AS)";
		def.description = "Clamps a value between min and max";
		def.category = "Math/Utility";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("value", "Value", "Value to clamp", DataType.F64).withDefaultF64(0.0));
		def.addPin(PinDefinition.input("min", "Min", "Minimum value", DataType.F64).withDefaultF64(0.0));
		def.addPin(PinDefinition.input("max", "Max", "Maximum value", DataType.F64).withDefaultF64(1.0));
		def.addPin(PinDefinition.output("exec_out", "Done", "Complete", DataType.Exec));
		def.addPin(PinDefinition.output("result", "Result", "Clamped value", DataType.F64));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		const value = ctx.getF64("value");
		const min = ctx.getF64("min");
		const max = ctx.getF64("max");
		ctx.setF64("result", Math.max(min, Math.min(max, value)));
		return ctx.success();
	}
}

const pkg = new NodePackage();
pkg.register(new AddNode());
pkg.register(new SubtractNode());
pkg.register(new MultiplyNode());
pkg.register(new DivideNode());
pkg.register(new ClampNode());

export function get_nodes(): i64 {
	return pkg.getNodes();
}

export function run(ptr: i32, len: i32): i64 {
	return pkg.run(ptr, len);
}
